### PR TITLE
[ONNX] Add dtype check in onnx verification

### LINKED ISSUE #78852: [ONNX] Some functions do not preserve shape for scalars
Discovered in #78701

We should use the comparison method in torch.testing instead of what's numpy to assert that shapes match

### PR DESCRIPTION
**Pull request by qqaatw**
Currently we don't have a dtype check in verifying the consistency between PyTorch and ONNX outputs. As a result, some of dtype inconsistencies were found and reported: #77842 #77845

This is a POC.

Failed workflows:
- [linux-xenial-py3.7-clang7-onnx / test (default, 2, 2, linux.2xlarge)]
  - inconsistent shape
    - TestONNXRuntime_opset10.test_all (#79371)
    - TestONNXRuntime_opset10.test_any (#79371)
    - TestONNXRuntime_opset10.test_argmin_argmax (#79503)
    - TestONNXRuntime_opset10.test_hardshrink (#79695)
    - TestONNXRuntime_opset10.test_linalg_norm (#79506)
    - TestONNXRuntime_opset10.test_linalg_vector_norm (#79506)
    - TestONNXRuntime_opset10.test_prelu_scalar (#79846)
    - TestONNXRuntime_opset10.test_softshrink (#79695)
    - TestONNXRuntime_opset10.test_sum_empty_tensor (skipped)
    - TestONNXRuntime_opset10.test_tolist (skipped)
  - inconsistent dtype
    - test_arithmetic_prim_bool (skipped)
    - test_arithmeticOps_with_low_precision (skipped)
    - test_arithmetic_prim_float (skipped)
    - test_logical_and (#79339)
    - test_logical_or (#79339)
    - test_logical_xor (#79339)
    - test_pow (skipped)
    - test_primitive_input_floating (skipped)
    - test_quantize_per_tensor (#79690)
    - test_quantized_adaptive_avg_pool2d (#79690)
    - test_quantized_arithmetic (#79690)
    - test_quantized_arithmetic_qfunctional (#79690)
    - test_quantized_conv2d (#79690)
    - test_quantized_conv2d_relu (#79690)
    - test_quantized_flatten (#79690)
    - test_quantized_hardsigmoid (#79690)
    - test_quantized_hardswish (#79690)
    - test_quantized_linear (#79690)
    - test_quantized_sigmoid (#79690)
    - test_item (skipped)
    - test_full_like_value (skipped)
    - TestONNXRuntime_opset7.test_div_rounding_mode (skipped)
    - TestONNXRuntime_opset8.test_div_rounding_mode (skipped)
    - TestONNXRuntime_opset9.test_div_rounding_mode (skipped)
    - TestONNXRuntime_opset9_IRv4.test_div_rounding_mode (skipped)
    - test_outer (skipped)
    - test_symbolic_shape_inference_arange_2 (skipped)

### CODE REVIEW

**Comment by justinchuby** [comment]
cc @BowenBao

**Comment by BowenBao** [comment]
@qqaatw This is awesome! Thank you so much for contribution!

A suggestion would be to add a new skip decorator to disable dtype checks for these tests. Then remove the skip in later follow up PRs that fix individual tests. This allows CI to tell us if the issue is properly fixed.

If you are interested you can utilize ghstack to submit follow-up individual fixes on top of this PR.

**Comment by qqaatw** [comment]
> @qqaatw This is awesome! Thank you so much for contribution!
> 
> A suggestion would be to add a new skip decorator to disable dtype checks for these tests. Then remove the skip in later follow up PRs that fix individual tests. This allows CI to tell us if the issue is properly fixed.
> 
> If you are interested you can utilize ghstack to submit follow-up individual fixes on top of this PR.

Thank you, good suggestion indeed!

ghstack requires write access to the repository, which I don't have now.

File: torch/onnx/verification.py [hunk 1]
```diff
@@ -113,14 +113,28 @@ def _ort_session(
     return ort_session
 
 
-def _compare_ort_pytorch_outputs(ort_outs, pt_outs, rtol, atol):
+def _compare_ort_pytorch_outputs(
+    ort_outs, pt_outs, rtol, atol, check_shape, check_dtype
```
Comment (justinchuby) [inline torch/onnx/verification.py:117]: Could you help type annotate this function? Thanks!

File: torch/onnx/verification.py [hunk 2]
```diff
@@ -113,14 +113,28 @@ def _ort_session(
...
 
-    assert len(pt_outs) == len(ort_outs), "number of outputs differ"
+    assert len(ort_outs) == len(
+        pt_outs
+    ), f"number of outputs differ ONNX runtime: ({len(ort_outs)}) PyTorch: ({len(pt_outs)})"
 
     for ort_out, pt_out in zip(ort_outs, pt_outs):
-        np.testing.assert_allclose(ort_out, pt_out, rtol=rtol, atol=atol)
+        if not check_shape:
+            # allows output shapes differ but are broadcastable.
```
Comment (justinchuby) [inline torch/onnx/verification.py:128]:
```suggestion
            # Allow different but broadcastable output shapes
```

**Comment by justinchuby** [comment]
> I was just about to reply the wrapper thing but found it was deleted :P

Yeah I confused myself. Could you use functools wrap on it?

**Comment by qqaatw** [comment]
> > I was just about to reply the wrapper thing but found it was deleted :P
> 
> Yeah I confused myself. Could you use functools wrap on it?

Done.

File: torch/onnx/verification.py [hunk 3]
```diff
@@ -113,14 +113,34 @@ def _ort_session(
     return ort_session
 
 
-def _compare_ort_pytorch_outputs(ort_outs, pt_outs, rtol, atol):
+def _compare_ort_pytorch_outputs(
+    ort_outs: List[np.ndarray],
+    pt_outs: List[torch.Tensor],
```
Comment (justinchuby) [inline torch/onnx/verification.py:118]: Sequence[...]

**Review by justinchuby — APPROVED** [review]
looking good. @BowenBao thoughts?

**Comment by ganler** [comment]
Hi @justinchuby, just curious, is this thread also partially inspired by https://github.com/pytorch/pytorch/pull/78701? Thanks!

**Review by BowenBao — APPROVED** [review]
LGTM, thanks!

**Comment by justinchuby** [comment]
Could you rebase?

**Comment by qqaatw** [comment]
> Could you rebase?

After current opened PRs are merged maybe?

**Pull request closed without merging**